File: generic-cpp-receiver/src/main/cpp/RCode/instreams/RCodeMarkerInStream.cpp
/*
 * RCodeMarkerInStream.cpp
 *
 *  Created on: 8 Dec 2020
 *      Author: robert
 */

#include "RCodeMarkerInStream.hpp"

char RCodeMarkerInStream::read() {
    if (hasRead) {
        sequenceIn->read();
    }
    hasRead = true;
    return sequenceIn->peek();
}

char RCodeMarkerInStream::reread() {
    return sequenceIn->peek();
}

bool RCodeMarkerInStream::hasNext() {
    return sequenceIn->hasNext();
}

void RCodeMarkerInStream::close() {
}

File: java-receiver/src/main/java/com/wittsfamily/rcode/javareceiver/instreams/RCodeMarkerInStream.java
package com.wittsfamily.rcode.javareceiver.instreams;

public class RCodeMarkerInStream {
    private final RCodeSequenceInStream sequenceIn;
    private boolean hasRead = false;

    public RCodeMarkerInStream(RCodeSequenceInStream sequenceIn) {
        this.sequenceIn = sequenceIn;
    }

    public void reset() {
        hasRead = false;
    }

    public char read() {
        if (hasRead) {
            sequenceIn.read();
        }
        hasRead = true;
        return sequenceIn.peek();
    }

    public char reread() {
        return sequenceIn.peek();
    }

    public void unread() {
        hasRead = false;
    }

    public boolean hasNext() {
        return sequenceIn.hasNext();
    }

    public void close() {
        sequenceIn.getCommandInStream().openWith(' ');
    }

    public void eatWhitespace() {
        char c = reread();
        while (c == ' ' || c == '\t' || c == '\r') {
            c = read();
        }
    }
}
